tests/spki_tests.rs

```diff
@@ -0,0 +1,701 @@
+extern crate ring;
+extern crate untrusted;
+extern crate rustc_serialize;
+
+use std::io::BufRead;
+use std::path::Path;
+use std::vec::Vec;
+use std::io::Read;
+use std::fs::File;
+use rustc_serialize::base64::FromBase64;
+use ring::der;
+use ring::signature::spki;
+use ring::signature::spki::VerifyWithSPKIError;
+#[cfg(feature = "rsa_signing")]
+use ring::signature;
+#[cfg(feature = "rsa_signing")]
+use ring::rand;
+
+// TODO: The expected results need to be modified for SHA-1 deprecation.
+
+macro_rules! test_verify_signature_pem {
+    ($fn_name:ident, $file_name:expr, $signature_alg:expr, $expected_result:expr) => {
+        #[test]
+        fn $fn_name() {
+            test_verify_signature_pem($file_name, $signature_alg, $expected_result);
+        }
+    }
+}
+
+fn test_verify_signature_pem(file_name: &str,
+                             signature_algorithm: &'static spki::Algorithm,
+                             expected_result: Result<(), VerifyWithSPKIError>) {
+    let tsd = parse_test_signed_data(file_name);
+    let spki_value = untrusted::Input::from(&tsd.spki);
+
+    let signature = untrusted::Input::from(&tsd.signature);
+    let signature = signature.read_all(VerifyWithSPKIError::BadDER, |input| {
+        der::bit_string_with_no_unused_bits(input)
+            .map_err(|_| VerifyWithSPKIError::BadDER)
+    }).unwrap();
+
+    assert_eq!(expected_result,
+    spki::verify(signature_algorithm,
+                 spki_value,
+                 untrusted::Input::from(&tsd.data),
+                 signature));
+}
+
+macro_rules! test_rsa_sign_verify_spki {
+    ($fn_name:ident, $signing_encoding:expr, $algorithm:expr, $key_id:expr) => {
+        #[cfg(feature = "rsa_signing")]
+        #[test]
+        fn $fn_name() {
+            rsa_sign_and_verify_spki($signing_encoding, $algorithm, $key_id);
+        }
+    }
+}
+
+macro_rules! test_rsa_verify_sig_file_spki {
```

Comment:
For the `sig_file` tests, instead of having macros and functions specific to RSA and ECDSA, let's refactor things so that we have ones that are generic to both. In particular, I think you can just rename the input files a bit so that they have “RSA_PKCS1_2048_8192_SHA256” and “ECDSA_P256_SHA256” in the name. Then you should be able to eliminate this code duplication because callers could pass `"RSA_PKCS1_2048_8192_SHA256"` instead of separate `"rsa_2048", "pkcs1", "sha256"` parameters, and they could pass `"ECDSA_P256_SHA256"` instead of separate `"ecdsa_prime256v1", "sha256"` parameters. In general, the *ring* API is designed to make it possible to write such generic code. If it's problematic to do this, let me know and I'll work through it in detail with you.